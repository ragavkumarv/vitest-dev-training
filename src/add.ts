// implement

function add(num1: number, num2: number) {
  if (typeof num1 != 'number' || typeof num2 != 'number') {
    throw new Error('Please provide valid number');
  }

  return num1 + num2;
}

export { add };
