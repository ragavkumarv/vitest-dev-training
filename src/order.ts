// src/utils/order.ts

export interface OrderItem {
  name: string;
  price: number;
  quantity: number;
  discount?: number; // Discount percentage (e.g., 10 for 10%)
}

export interface TaxRule {
  rate: number; // Tax rate percentage (e.g., 5 for 5%)
  applyToDiscountedPrice?: boolean; // Whether to apply tax after discount
}

/**
 * Validates the properties of an OrderItem.
 * @param item - The order item to validate.
 */
export function validateOrderItem(item: OrderItem): void {
  const { name, quantity, discount } = item;

  if (quantity < 0) {
    throw new Error(`Quantity for item "${name}" cannot be negative`);
  }

  if (discount !== undefined && (discount < 0 || discount > 100)) {
    throw new Error(`Invalid discount for item "${name}"`);
  }
}

/**
 * Calculates the total price for a single item after applying discount.
 * @param item - The order item.
 * @returns The total price for the item.
 */
export function calculateItemTotal(item: OrderItem): number {
  validateOrderItem(item);

  const { price, quantity, discount = 0 } = item;
  const totalPrice = price * quantity;
  const discountAmount = (totalPrice * discount) / 100;

  return totalPrice - discountAmount;
}

/**
 * Calculates the subtotal for all items.
 * @param items - Array of order items.
 * @returns The subtotal amount.
 */
export function calculateSubtotal(items: OrderItem[]): number {
  return items.reduce(
    (subtotal, item) => subtotal + calculateItemTotal(item),
    0
  );
}

/**
 * Calculates the tax amount based on the tax rule.
 * @param subtotal - The subtotal amount.
 * @param items - Array of order items.
 * @param taxRule - The tax rule to apply.
 * @returns The tax amount.
 */
export function calculateTax(
  subtotal: number,
  items: OrderItem[],
  taxRule: TaxRule
): number {
  const { rate, applyToDiscountedPrice = true } = taxRule;

  if (rate <= 0) return 0;

  const baseAmount = applyToDiscountedPrice
    ? subtotal
    : items.reduce((sum, { price, quantity }) => sum + price * quantity, 0);

  return (baseAmount * rate) / 100;
}

/**
 * Calculates the total order amount, including taxes.
 * @param items - Array of order items.
 * @param taxRule - The tax rule to apply.
 * @returns The total order amount.
 */
export function calculateOrderTotal(
  items: OrderItem[],
  taxRule: TaxRule
): number {
  if (!items.length) return 0;

  const subtotal = calculateSubtotal(items);
  const tax = calculateTax(subtotal, items, taxRule);
  const total = subtotal + tax;

  return parseFloat(total.toFixed(2));
}
