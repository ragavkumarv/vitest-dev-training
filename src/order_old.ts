export interface OrderItem {
  name: string;
  price: number;
  quantity: number;
  discount?: number; // Discount in percentage (e.g., 10 for 10%)
}

export interface TaxRule {
  rate: number; // Tax rate in percentage (e.g., 5 for 5%)
  applyToDiscountedPrice?: boolean; // Whether to apply tax after discount
}

export function calculateOrderTotal(
  items: OrderItem[],
  taxRule: TaxRule
): number {
  if (!items.length) return 0;

  let subtotal = 0;

  for (const item of items) {
    if (item.quantity < 0) {
      throw new Error(`Quantity for item ${item.name} cannot be negative`);
    }

    let itemTotal = item.price * item.quantity;

    if (item.discount) {
      if (item.discount < 0 || item.discount > 100) {
        throw new Error(`Invalid discount for item ${item.name}`);
      }
      itemTotal -= (itemTotal * item.discount) / 100;
    }

    subtotal += itemTotal;
  }

  let total = subtotal;

  if (taxRule.rate > 0) {
    if (taxRule.applyToDiscountedPrice) {
      total += (subtotal * taxRule.rate) / 100;
    } else {
      const originalTotal = items.reduce(
        (sum, item) => sum + item.price * item.quantity,
        0
      );
      total += (originalTotal * taxRule.rate) / 100;
    }
  }

  return parseFloat(total.toFixed(2));
}
