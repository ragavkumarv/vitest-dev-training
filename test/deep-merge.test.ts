import { assert, describe, expect, test, it } from 'vitest';
import { deepMerge } from '../src/deep-merge';

describe('Merge in two elements', () => {
  it('Merging two arrays', () => {
    const merged = deepMerge(
      ['apple 🍎', 'orange 🍊'],
      ['peach 🍑', 'cocunut 🥥']
    );

    expect(merged).toEqual(['apple 🍎', 'orange 🍊', 'peach 🍑', 'cocunut 🥥']);
  });

  it('Merging two objects', () => {
    const merged = deepMerge(
      { name: 'guru', github: 'unknown' },
      { github: 'Gnanaguru18', age: 20 }
    );

    expect(merged).toEqual({
      age: 20,
      github: 'Gnanaguru18',
      name: 'guru',
    });
  });

  it('Deep Merging two nested objects', () => {
    const merged = deepMerge(
      { name: 'guru', age: 20, accounts: { github: 'unknown' } },
      { accounts: { twitter: 'Gnanaguru18' } }
    );

    expect(merged).toEqual({
      age: 20,
      accounts: { github: 'unknown', twitter: 'Gnanaguru18' },
      name: 'guru',
    });
  });
});

// toBe -> primitive
// toEqual -> objects, arrays, nested
