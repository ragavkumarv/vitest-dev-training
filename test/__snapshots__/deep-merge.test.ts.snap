// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Merge in two elements > Deep Merging two nested objects - Snapshot 1`] = `
{
  "accounts": {
    "github": "unknown",
    "twitter": "Gnanaguru18",
  },
  "age": 20,
  "name": "guru",
}
`;
